Proof of Indexing

That3Percent opened the pull request:
### Proof of Indexing
This change adds much of the needed framework around the Proof of Indexing digest update and serving. Some of the database work is left for future work in order to integrate what changes have been marked. These are clearly marked.

Changes were made to the network RFC in order to work around limitations of GraphQL. In particular, interfaces cannot be used as inputs.

### Other new APIs
new `object!` macro - removes noise/boilerplate for graphql objects (`IntoValue` added to support this). Lib code has been migrated to use this, There are more than 100 places within tests to migrate later.
`Store::with_entity_connection` - Talk to the database without violating the futures API, allowing timeout, quicker transaction cancellation on future drop, and greater levels of concurrency
`TryFrom<BigInt> for u64` - fallible conversion with custom error
`trait CancelToken` - convenient `token.check_cancel()?` API.
More `CheapClone` impls
`spawn_blocking_async_allow_panic` to eventually replace futures based version

### Fixes
Call `resolve_scalar_value` when outside of an object for GraphQL
No 6 hour timeout for db connection in test means you get an error message when the connection fails

### And more
Various added documentation, migrating to async, and other cleanup

lutter commented on store/postgres/src/relational.rs:
```diff
@@ -283,6 +283,30 @@ impl Layout {
+                artifact bigint not null
+            );",
+            proof_of_indexing_name,
+        )?;
```
Not a review, just some thoughts since last we talked about this:

(1) A POI artifact is likely valid for a block range (from the block where the last change happened to just before the block with the next change) That means artifacts should store a `block_range`, and makes them really just another entity type. We should just inject the appropriate `Table` definition into the `Layout`'s list of tables and then manage them with the usual entity methods. 

(2) We should hold off on any marking of artifacts by `data_source` since we don't need it yet. For now, it's enough to have one series of artifacts since each subgraph has a single source of time. We should wait with adding a `data_source` or similar field until we actually implement something where a subgraph has more than one source of time.

That3Percent replied:
@lutter: 
Using the entity API turned out [nicely](https://github.com/graphprotocol/graph-node/blob/df2dc6b9bc31cb3246c4d2661e25ea1a0496b20f/core/src/subgraph/instance_manager.rs#L824):

 Is this what you had in mind for determining [whether the POI table exists](https://github.com/graphprotocol/graph-node/blob/df2dc6b9bc31cb3246c4d2661e25ea1a0496b20f/core/src/subgraph/instance_manager.rs#L811)? I'm unclear as to whether adding this to the `Schema` for the side effect of having the tables created and be queryable is the right approach, but didn't see any straightforward alternatives. Where would you put the code that edits the `Schema` to include the additional type definition?

lutter replied:
> @lutter:
> Using the entity API turned out [nicely](https://github.com/graphprotocol/graph-node/blob/df2dc6b9bc31cb3246c4d2661e25ea1a0496b20f/core/src/subgraph/instance_manager.rs#L824):

Nice! You can't use `$poi` as the entity type though: Postgres names can contain a `$`, but can't start with it [(docs)](https://www.postgresql.org/docs/12/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS).

> Is this what you had in mind for determining [whether the POI table exists](https://github.com/graphprotocol/graph-node/blob/df2dc6b9bc31cb3246c4d2661e25ea1a0496b20f/core/src/subgraph/instance_manager.rs#L811)? I'm unclear as to whether adding this to the `Schema` for the side effect of having the tables created and be queryable is the right approach, but didn't see any straightforward alternatives. Where would you put the code that edits the `Schema` to include the additional type definition?

What I had in mind there was a check against the actual database schema; externally, that should be a method `Store.supports_poi(&subgraph_id)` which is backed by the same method on `entities::Connection` which in turn always returns `false` for `JsonStorage` and for `Storage::Relational` queries [`information_schema.tables`](https://www.postgresql.org/docs/9.6/infoschema-tables.html) to see if an entry with  `table_name = <name of POI table> and table_schema = <layout.schema>` exists. The result of that query should be cached in the `Layout` and so should probably be passed into `Layout::new` (and `struct Layout` gets a new attribute `supports_poi`)

That3Percent commented:
This is now fully operational.

lutter commented on store/postgres/src/store.rs:
```diff
@@ -665,17 +704,116 @@ impl Store {
+    ///   when the supplied cancel token is not cancelled.
+    // Ideally, we want to pass a borrow of the PooledConnection in. But,
+    // with_entity_conn requires this to be owned to give it to e::Connection
+    // If e::Connection could take a reference this would not be a problem
```
What would it take to make it so that we can pass a reference of `PooledConnection` in?

That3Percent replied:
`entities::Connection`` would need to have a lifetime on it. This seems like the right thing to do. Let me do a quick dive and see how much lift it would require.

lutter commented on store/postgres/src/store.rs:
```diff
@@ -665,17 +704,116 @@ impl Store {
+    // Ideally, we want to pass a borrow of the PooledConnection in. But,
+    // with_entity_conn requires this to be owned to give it to e::Connection
+    // If e::Connection could take a reference this would not be a problem
+    async fn with_conn<T: Send + 'static>(
```
This function and `with_entity_conn` is fundamental enough that it deserves documentation on how to use it properly, in particular, for which operations to use it and for which ones not to use it and instead fall back to `Store::get_conn`.

Your uses call `cancel_handle.check_cancel()?` as the first thing - is that necessary seeing how `with_conn` calls that right before calling `f`. As a user of this, when should I be calling  `cancel_handle.check_cancel()?` ?

That3Percent replied:
You're right. I'll add copious amounts of documentation around `with_conn` and `with_entity_conn` that encompasses a lot of the earlier discussions.

I don't think there is any case where one should fall back on `Store::get_conn`. The one place where it might make seem to make sense _right now_ is where you need multiple connections simultaneously (eg: a normal connection and an entity connection, or multiple entity connections) when you're already inside `with_conn`. But, there is a better way to deal with that for a slightly different API. Instead of using `with_entity_conn` as it exists today, simply have `with_conn` return some kind of wrapper type that has a method `to_entity_conn<'a>(&'a self) -> e::Connection<'a>` which would allow for re-use of the same connection multiple times in the same `with_conn` closure.

I wish I had time to port the rest of the database code and refine the API, but I think this is going to have to wait.

That3Percent replied:
Screw it. Maybe I'll go ahead and make that API since now having the good API doesn't take long, it's not inferring that everything needs to be migrated.

That3Percent replied:
Discussed offline, will not make these changes yet.

lutter commented on store/postgres/src/store.rs:
```diff
@@ -665,17 +704,116 @@ impl Store {
-            warn!(self.logger, "Possible contention in DB connection pool";
-                               "wait_ms" => wait.as_millis())
+            warn!(&self.logger, "Possible contention in DB connection pool";
+                                "wait_ms" => wait.as_millis())
```
I'd be hugely in favor of just getting rid of this log statement. I don't think we get much utility out of it anymore.

lutter commented on store/postgres/src/store.rs:
```diff
@@ -854,6 +993,69 @@ impl StoreTrait for Store {
+            let entities = conn
+                .query(
+                    &logger,
+                    EntityCollection::All(vec!["poi_artifact".to_owned()]),
```
Using a string constant here would be very good (and this one is actually the entity type, i.e., `Table.object`)

lutter commented on store/postgres/src/relational.rs:
```diff
@@ -170,8 +168,43 @@ impl Layout {
+                /// is really only needed for the tests to make the names of indexes
+                /// predictable
+                position: tables.len() as u32,
+            };
```
How fun, a manually constructed `Table` structure :laughing:

That3Percent replied:
Was there a better way to do this? I didn't want to create a schema type and pass that to `Table::new`.

lutter replied:
No, that's perfectly fine and probably the best you can do here w/o huge overhead

lutter commented on runtime/wasm/src/host_exports.rs:
```diff
@@ -129,13 +130,30 @@ impl HostExports {
+    /// and merge the results later. Right now, this is just ethereum but it
+    /// will be expanded for ipfs and the availability chain.
+    fn poi_namespace(&self) -> &str {
+        "ethereum"
```
Strictly speaking, this should be the concrete network, not just the family. For now, it doesn't matter since we don't ever have subgraphs that straddle networks, so that for a mainnet subgraph `ethereum` means `mainnet` and for a Rinkeby subgraph it means `rinkeby`, but it would be cleaner to actually use the network.

Jannis replied:
Yeah, agreed

lutter commented on graphql/src/values/mod.rs:
```diff
@@ -1,15 +1,77 @@
+
+/// Creates a `graphql_parser::query::Value::Object` from key/value pairs.
+#[macro_export]
+macro_rules! object {
```
Neat! That's a nice tool to tighten up some of our code. At least for our tests, would also be nice to have an `entity!` macro along the same lines.

That3Percent replied:
I'd be happy to create one. Can you point me to one test that would use this macro so I know what you're looking for?

That3Percent replied:
Nevermind, I think you mean just the search for `Entity::new`

lutter replied:
It's actually pretty easy since `graph::data::store::Value` already implements `From` for a lot of cases:
```rust
macro_rules! entity {
    ($($name:ident: $value:expr,)*) => {
        {
            let mut result = ::graph::prelude::Entity::new();
            $(
                result.insert(stringify!($name).to_string(), Value::from($value));
            )*
            result
        }
    };
    ($($name:ident: $value:expr),*) => {
        entity! {$($name: $value,)*}
    };
}
```

Jannis commented on server/index-node/src/resolver.rs:
```diff
@@ -34,19 +35,41 @@ impl TryFromValue for DeploymentAssignment {
 
+#[derive(Debug)]
+struct ProofOfIndexingInput {
+    // The id of the subgraph.
+    subgraph_name: String,
```
This should really be the ID (i.e. the `Qm...` hash of the subgraph deployment), not the subgraph name. The name is a moving target, whereas the ID always points to the same subgraph data.

Jannis commented on server/index-node/src/resolver.rs:
```diff
@@ -34,19 +35,41 @@ impl TryFromValue for DeploymentAssignment {
+    // The block number in ethereum to query against. Eventually when the data
+    // availability chain comes online and we support other block chains then
+    // this will be come an optional field.
+    ethereum_block_number: BigInt,
```
IMHO this should be just `block` or `block_number` (with the comment reflecting that it's the block number on the chain that the subgraph is indexing, which could be an Ethereum chain). The reason I'd like to get rid of the `ethereum` here is that this will work just as well for other blockchains.

That3Percent replied:
We had started out with an API in the RFC that was something like `blocks: [Block!]!` where `Block` was an interface and there was `EthereumBlock` etc. That doesn't work, because GraphQL doesn't (yet?) support interfaces in inputs. We still need to be able to support blocks from multiple blockchains in the same query - especially with the availability chain coming.

So, there's a couple of APIs possible. One, you could keep in the spirit of the per/network blocks without the interface, but then it's easy to specify more than one block with the same network. Eg: `[{Block {network: "eth", number: 10}, Block {network: "eth", number: 20}}]` It's unclear what to return (probably an error, even though the data is not malformed). With the current API, this isn't a problem and has the same amount of API churn as the original `EthereumBlock` API.

I'm not really happy with either API really. Maybe we can defer until we actually have multiple chains?

Jannis commented on server/index-node/src/resolver.rs:
```diff
@@ -470,17 +469,133 @@ where
+            Ok(Some(poi)) => q::Value::String(poi.0),
+            Ok(None) => q::Value::Null,
+            Err(e) => {
+                error!(
+                    self.logger,
+                    "Failed to query proof of indexing";
+                    "subgraph" => subgraph_name,
+                    "error" => format!("{:?}", e)
+                );
```
Wouldn't we want to return the error to the client?

That3Percent replied:
I would have thought so, but this is the pattern that other resolvers in the file are following.

Jannis commented on server/index-node/src/schema.graphql:
```diff
@@ -7,6 +7,12 @@ scalar String
 type Query {
   indexingStatusesForSubgraphName(subgraphName: String!): [SubgraphIndexingStatus!]!
   indexingStatuses(subgraphs: [String!]): [SubgraphIndexingStatus!]!
+  proofOfIndexing(request: ProofOfIndexingInput!): Bytes
```
In the interest of future-proofing, should this perhaps return an array of proof-of-indexing "slices"/"units"/"segments"... (whatever we call them)? Or would we combine them on the node before returning the result?

That3Percent replied:
This already combines them, which for The Network at least is already the desired outcome. If we want to output intermediate values for integration tests or something we could create a new API for that. I think we should leave that as a future improvement for now.

Jannis commented on server/index-node/src/schema.graphql:
```diff
@@ -7,6 +7,12 @@ scalar String
+
+input ProofOfIndexingInput {
+  subgraphName: String!
+  ethereumBlockNumber: BigInt!
```
block` or `blockNumber`

Jannis commented on server/index-node/src/schema.graphql:
```diff
@@ -7,6 +7,12 @@ scalar String
+}
+
+input ProofOfIndexingInput {
+  subgraphName: String!
```
`subgraph` is fine, but it should be the ID of the subgraph deployment, not the name of a subgraph.

Jannis commented on server/index-node/src/resolver.rs:
```diff
@@ -470,17 +469,133 @@ where
+            ))),
+        };
+
+        // Execute the query
+        let result = self
+            .graphql_runner
+            .run_query_with_complexity(query, None, None, Some(std::u32::MAX))
+            .wait()
+            .expect("error querying subgraph deployments");
+
+        let data = match result.data {
+            Some(data) => data,
+            None => {
+                error!(
+                    self.logger,
+                    "Failed to query subgraph deployments";
+                    "subgraph" => subgraph_name,
+                    "errors" => format!("{:?}", result.errors)
+                );
+                return Ok(q::Value::Null);
+            }
+        };
+
+        let subgraphs = match data
+            .get_optional::<q::Value>("subgraphs")
+            .expect("invalid subgraphs")
+        {
+            Some(subgraphs) => subgraphs,
+            None => return Ok(q::Value::Null),
+        };
+
+        let mut subgraphs = subgraphs
+            .get_values::<q::Value>()
+            .expect("invalid subgraph values");
+
+        let subgraph = if let Some(subgraph) = subgraphs.pop() {
+            subgraph
+        } else {
+            return Ok(q::Value::Null);
+        };
```
Most of this we can get rid of when we query by subgraph ID.

That3Percent replied:
I changed it to subgraph id, but unless I'm misunderstanding all this code stays. We still need to go from a subgraph id to a deployment id.

Jannis replied:
Subgraph ID = deployment ID 😉

Jannis commented on store/postgres/src/entities.rs:
```diff
@@ -823,6 +832,28 @@ fn find_schema(
         .optional()?)
 }
 
+fn supports_proof_of_indexing(
+    conn: &diesel::pg::PgConnection,
+    subgraph_id: &SubgraphDeploymentId,
+    schema: &str,
+) -> Result<bool, StoreError> {
+    if subgraph_id == &*SUBGRAPHS_ID {
+        return Ok(false);
+    }
+    #[derive(Debug, QueryableByName)]
+    struct Table {
+        #[sql_type = "Text"]
+        pub table_name: String,
+    }
+    let query =
+        "SELECT table_name FROM information_schema.tables WHERE table_schema=$1 AND table_name=$2";
+    let result: Vec<Table> = diesel::sql_query(query)
+        .bind::<Text, _>(schema)
+        .bind::<Text, _>("poi_artifact")
+        .load(conn)?;
+    Ok(result.len() > 0)
+}
+
```
I wonder if that should go into the `metadata` module that @lutter is preparing in the grafting PR?

lutter replied:
Yeah, that would be a good place for it (this query could also be written using Diesel directly, but I don't want @That3Percent have to go through the whole `table!` etc. dance I can fix that up once this is committed. We should just remember to do that.

Jannis commented on store/postgres/src/entities.rs:
```diff
@@ -1291,11 +1322,13 @@ impl Storage {
             }
             V::Relational => {
                 let subgraph_schema = store.input_schema(subgraph)?;
+                let create_poi = supports_proof_of_indexing(conn, subgraph, &schema.name)?;
```
`supports_poi` sounds good to me.

Jannis commented on store/postgres/src/relational.rs:
```diff
@@ -170,8 +168,43 @@ impl Layout {
-            match defn {
+        if create_proof_of_indexing {
+            let poi_table = Table {
+                object: "poi_artifact".to_owned(),
```
This type (and, more importantly, the generated table) must not collide with subgraph-defined entities. I therefor suggest we use `_PoiArtifact` (or `_POIArtifact`) for the entity type name and `_poi_artifact` for the table name.

That3Percent commented on core/src/subgraph/instance_manager.rs:
```diff
@@ -767,6 +778,62 @@ where
+    deployment_id: &SubgraphDeploymentId,
+    entity_cache: &mut EntityCache,
+) -> Result<(), Error> {
+    // Need to take this out whether or not we hit the early return. Otherwise it accumulates
+    let mut proof_of_indexing = if let Some(proof_of_indexing) = proof_of_indexing.take() {
+        proof_of_indexing
+    } else {
+        return Ok(());
+    };
```
It takes any poi changes out of the block state to be converted to entities.

That3Percent commented on core/src/subgraph/instance_manager.rs:
```diff
@@ -767,6 +778,62 @@ where
+
+    for (namespace, stream) in proof_of_indexing.drain() {
+        // Create the special POI entity key specific to this namespace
+        let entity_key = EntityKey {
+            subgraph_id: deployment_id.clone(),
+            entity_type: "poi_artifact".to_owned(),
+            entity_id: namespace,
+        };
```
I went down this road per your suggestion but the `TypedEntity`s aren't per-subgraph, they have the `SUBGRAPHS_ID` hardcoded in. That could be changed but it seems like more trouble than it's worth just to not construct an `EntityKey` here.

That3Percent commented on graph/src/components/store.rs:
```diff
@@ -1194,6 +1210,136 @@ pub trait Store: Send + Sync + 'static {
+    }
+}
+
+// The store trait must be implemented manually because mockall does not support async_trait, nor borrowing from arguments.
```
`DynTryFuture<'a, T>` is enough to create a problem for mockall.

That3Percent commented on graph/src/components/subgraph/proof_of_indexing.rs:
```diff
@@ -0,0 +1,139 @@
+/// The POI is the StableHash of:
+/// (Vec<ProofOfIndexingEvent, PreviousArtifact>)
+/// This struct contains the necessary state to construct that value in a streaming manner
+pub struct ProofOfIndexingStream {
```
The term digest refers to the output of a hash function. So, I'm renaming all cases of "artifact" to "digest" at the moment. For consistency, I think I'll call this `ProofOfIndexingHashFn`.

That3Percent commented on graph/src/components/subgraph/proof_of_indexing.rs:
```diff
@@ -0,0 +1,139 @@
+
+/// An intermediate value of a ProofOfIndexing specific to the events emitted by a single DataSource
+#[derive(Debug)]
+pub struct ProofOfIndexingArtifact(pub String);
```
The comment above this was outdated. This is now going to be `ProofOfIndexingDigest` - but it refers to both intermediate digests and final digests, which don't really need to be disambiguated by the type system.

That3Percent commented on graph/src/components/subgraph/proof_of_indexing.rs:
```diff
@@ -0,0 +1,139 @@
+    /// some data sources (eg: IPFS files) may be unreliable and therefore cannot mix
+    /// state with other data sources. This may also give us some freedom to change
+    /// the order of triggers in the future.
+    per_namespace: HashMap<String, ProofOfIndexingStream>,
```
I've renamed these to "causality_region"

That3Percent commented on graph/src/components/subgraph/proof_of_indexing.rs:
```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    fn emit(&mut self, event: &ProofOfIndexingEvent) {
```
I've renamed this to "write"

That3Percent commented on graph/src/ext/futures.rs:
```diff
@@ -97,8 +97,30 @@ pub struct CancelHandle {
+    }
+}
+
+pub struct Canceled;
```
`CancelableError` is a generic type, so that would cause `check_cancel` to also be generic, but then type inference wouldn't know what type `t.check_cancel()?;` should return - `CancelableError<???>`. This way fixes that issue, but also allows `Canceled` to coerce to other types. For example you could sometimes just have `Result<T, Canceled>` if the operation was otherwise infallible.

That3Percent commented:
@Jannis @lutter I think I've addressed all the comments, though some of them I'm not sure there was a satisfactory answer for.

Here is a summary of the latest changes:

- Query POI by subgraph ID, not name
- `with_conn` closure gets a reference to the PgConnection
- Named constants for duplicated string values
- Renamed `create_poi` to `has_poi`
- Renamed "artifact" to "digest"
- Add lots of documentation to `with_conn` and `with_entity_conn`
- Removed logs that are no longer providing value
- Renamed "namespace" to "causality_region"
- Include Ethereum network in causality region
- Figured out how to use GraphQL illegal names in PoI table names
- Added `entity!` and migrated a small handful of places in the code to use it
- Fixed unfinished comment
- Renamed `emit` to `write`

I think everything that wasn't fixed has some comment or question in the discussion here on GitHub.

Jannis commented on store/postgres/src/store.rs:
```diff
@@ -854,6 +1019,69 @@ impl StoreTrait for Store {
+                })
+                .collect::<Result<HashMap<_, _>, Error>>()?;
+
+            let hash = stable_hash_with_hasher::<twox_hash::XxHash64, _>(&by_causality_region);
```
I checked stable-hash to understand how `HashMap` (by definition unordered) hashes deterministically. I still don't understand it. Could you once more explain how that's guaranteed? I see there are tests for `HashMap` in stable-hash, so this is just for me to understand. 😉

That3Percent replied:
The properties we _must_ demonstrate are:

- Determinism: Any value which `PartialEq(a, b) == true` outputs `StableHash(a) == StableHash(b)`.
- Avalanche: Small changes in the input drastically change the output "seemingly randomly"
- Collision resistance: It is not easy to have 2 inputs that hash to the same value.

The algorithm for an unordered collection of *unique* (<- very important to be unique) items is as follows:
```
1: Initialize some binary value (rollup) to all 0 (or some predefined value), and count to 0
2: For each element in the collection:
    a: StableHash it independently of the supplied hash state
    b: XOR the intermediate digest from step a with the rollup
    c: increment count
3: Feed rollup into the StableHash
4: Feed the count into the StableHash
```

### Determinism
We'll take as a given that the StableHash of each element in the sequence has a deterministic stable hash.
Consider the rollup value as an array of bits. Each XOR which operates on the rollup bits operates on each bit independently. Eg: rollup_bits[0] is a function of all of the intermediate_digest_bits[0], and is not a function of any bits[1] or intermediate_digest_bits[1] etc.
Each XOR flips the rollup bit from 0 -> 1 or back if and only if digest_bit is 1. So, the final value of the bit is `count_ones(digest bits) % 2`. The function `count_ones(digest bits)` is independent of the order of `digest bits` and therefore independent of the order of the sequence.
Showing that `rollup` is deterministic was the hard part, it's pretty easy to see that `count` and what we do after is deterministic so that's left as an exercise for the reader.

### Avalanche
This falls out of 2 properties - first that each individual item in the sequence has it's own `StableHash` with the avalanche property, and secondly that XOR acts as a one-time-pad cipher with the probability of each bit being exactly 50%.

### Collision resistance
This falls out of each `StableHash` value being pseudo-unique, there not being duplicate items in the sequence, and including the count.

If there are duplicate items then a hash of `ABA` is the same as the hash of `B` because the XOR cancels.

The count is necessary for any implementation of `StableHasher` which might output `0` for the empty sequence. Such a stable hasher would make it so that `HashSet([])` and `HashSet([0])` collide. I guess this probably isn't necessary in practice.

### Performance

Another property we would _like_ to show is we can get good performance or that the chosen algorithm scales well for performance as compared to other more obvious methods.

The other "obvious" method would have been to sort the sequence and feed the sorted sequence through the hasher.

Sorting takes `O(N*log(N))` instead of `N` but more crucially requires that all the data be loaded and implement `Ord` (ruling out `f64` keys, among others). Although the implementation doesn't make use of this yet, not sorting makes it possible to stream batches of ipfs data sources from the database and update the hash incrementally.

leoyvens replied:
While that's a great explanation, I consider using `HashMap` tricky in code where determinism is a requirement. It's easy to iterate over the items without remembering that the order is not deterministic. So we should prefer `BTreeMap` whenever possible, it would be nice to have a `StableHash` implementation for it.

That3Percent replied:
@leoyvens It would be nice to have a `StableHash` implementation for `BTreeMap`. I'm not convinced about preferring `BTreeMap` in the general case though. [The docs](https://doc.rust-lang.org/std/collections/index.html#use-a-hashmap-when) have some nice guidelines about when to choose `HashMap` or `BTreeMap` which I find agreeable.

leoyvens replied:
Those docs kinda suggest defaulting to `HashMap` if you don't need `BTreeMap`, but I feel the opposite. Not all `HashMap` operations are deterministic. All `BTreeMap` operations are deterministic. In any codebase avoiding accidental non-determinism is good, and in ours I feel `HashMap` is a needless risk.

leoyvens replied:
Anyways I'm not saying this PR needs to change, it's just a general note.

That3Percent replied:
"any codebase avoiding accidental non-determinism" -> almost all codebases? I mean, it's not like our application is unusual in requiring determinism. So, you're pretty much condemning the use of `HashMap` almost everywhere.

In my experience, this is a bug that bites every developer... _once_. And it's usually a pretty easy fix with consequences not larger than other ordinary bugs. When we consider broadly applying a set of rules as a preventative measure we need to take into account the tradeoffs against the likelihood and severity of the problem manifesting.

It's a pretty subtle issue that I think reasonable people can disagree on.  I'd still choose `HashMap` here again. But, I certainly appreciate the sentiment from which your concern emerges and can respect that you may choose `BTreeMap`.

That3Percent commented on server/index-node/src/schema.graphql:
```diff
@@ -7,6 +7,12 @@ scalar String
 type Query {
   indexingStatusesForSubgraphName(subgraphName: String!): [SubgraphIndexingStatus!]!
   indexingStatuses(subgraphs: [String!]): [SubgraphIndexingStatus!]!
+  proofOfIndexing(request: ProofOfIndexingInput!): Bytes
```
It feels more natural _now_, but when there are some arbitrary count of block numbers because of subgraph composition over the availability chain and other non-Ethereum chains it's going to be weird that only the Ethereum chain is special in that it doesn't have a prefix for it's block number, and there are a bunch of optional arguments rather than optional fields on a struct.

I guess the future can wait though - especially since this is an internal API. Kind of hesitant about putting effort into changing it though.

Jannis replied:
Yeah, understood. We can leave it the way it is.

That3Percent replied:
It's ok, I already changed it to your preference. This makes it the best code according to the current requirements.

That3Percent commented on store/postgres/src/entities.rs:
```diff
@@ -823,6 +834,28 @@ fn find_schema(
         .optional()?)
 }
 
+fn supports_proof_of_indexing(
```
Yeah, I don't know either. It does work though... and only in that specific combination.

That3Percent commented on store/postgres/src/store.rs:
```diff
@@ -665,17 +704,142 @@ impl Store {
+    /// Execute a closure with a connection to the database.
+    ///
+    /// # API
+    ///   The API of using a closure to wrap the use of wrapper serves several
```
Sometimes I do this thing where I type a sentence and then start thinking of something else and then dont

That3Percent commented on store/postgres/src/store.rs:
```diff
@@ -665,17 +704,142 @@ impl Store {
+        match result {
+            Ok(t) => Ok(t),
+            Err(CancelableError::Error(e)) => Err(e),
+            Err(CancelableError::Cancel) => panic!("The closure supplied to with_entity_conn must not return Err(Canceled) unless the supplied token was canceled."),
```
The cancel token supplied to the task closure is only canceled if this future is dropped while we were waiting for the task to finish. If we reached this point, the supplied closure must have finished executing before the cancel token was canceled. If the value is then `CancelableError::Cancel` the method decided to return `CancelableError::Cancel` even though cancellation was not requested. In that case, there is no good return value for `with_conn` since it does not return `CancelableError` (nor should it, since it follows the better drop-to-cancel pattern).
